### PR TITLE
Metadata-based state & command description providers

### PR DESCRIPTION
**Pull request by ghys**
Implements #1185.

These providers will look into item metadata, which
can be managed by UIs with the API, to set or override
the item's state description (pattern, options, read
only...) or command description.

Signed-off-by: Yannick Schaus <github@schaus.net>

### CODE REVIEW

File: bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataCommandDescriptionProvider.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            } catch (Exception e) {
+                logger.error("Unable to parse the commandDescription from metadata for item {}", itemName, e);
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    @Reference
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataCommandDescriptionProvider.java]: You can add the `@Reference` directly in line 45 and get rid off the getter and setter methods.
Or pass it in via constructor as you do in `MetadataStateDescriptionFragmentProvider`.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                                return (o.contains("="))
+                                        ? new StateOption(o.split("=")[0].trim(), o.split("=")[1].trim())
+                                        : new StateOption(o.trim(), null);
+                            }).collect(Collectors.toList());
+                    builder.withOptions(stateOptions);
+                }
+
+                return builder.build();
+            } catch (Exception e) {
+                logger.error("Unable to parse the stateDescription from metadata for item {}", itemName, e);
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java]: This should imho be a WARN for the user, telling him that the value is ignored. No stacktrace must be added here as we are not talking about a bug in the software.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataCommandDescriptionProvider.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                                    new CommandOption(o.split("=")[0].trim(), o.split("=")[1].trim()));
+                        } else {
+                            commandDescription.addCommandOption(new CommandOption(o.trim(), null));
+                        }
+                    });
+                }
+
+                return commandDescription;
+            } catch (Exception e) {
+                logger.error("Unable to parse the commandDescription from metadata for item {}", itemName, e);
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataCommandDescriptionProvider.java]: This should imho be a WARN for the user, telling him that the value is ignored. No stacktrace must be added here as we are not talking about a bug in the software.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java [hunk 4]
```diff
@@ -83,10 +83,10 @@ public MetadataStateDescriptionFragmentProvider(final @Reference MetadataRegistr
                     builder.withMinimum(getBigDecimal(metadata.getConfiguration().get("min")));
                 }
                 if (metadata.getConfiguration().containsKey("max")) {
-                    builder.withMaximum(getBigDecimal(metadata.getConfiguration().get("min")));
+                    builder.withMaximum(getBigDecimal(metadata.getConfiguration().get("max")));
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java]: haha, it seems the unit tests have already helped ;-)

**Comment by kaikreuzer** [comment]
Now you only have to convince Jenkins:
```
07:44:24 [ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/bundles/org.openhab.core/pom.xml [0:0]: Constructors must be public access.
07:44:24 [ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/bundles/org.openhab.core/pom.xml [0:0]: In component org.openhab.core.internal.items.MetadataCommandDescriptionProvider, @Reference can only be used for parameters on the constructor annotated @Activate
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java (hunk 4 above)
Comment (ghys) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java]: My thoughts exactly but I wanted to keep that under the radar ;)

**Pull request merged**